### DISCUSSION TITLE
Alguém pode me explicar como é resolvido essa problema? Grato!

### DISCUSSION THREAD

**Discussion by LuizFelipe3** [frontendbr/forum#1979] [Perguntas e Respostas]
Observe o código abaixo,  que ilustra o uso de uma função recursiva.
private int a, b, c;
private int calc(int p)
{
c += 1;
if (p < 2)
{
return 4;
}
else
{
return c + calc(p - 1);
}
}
void main()
{
b = 3;
c = 0;
a = calc(b);
Console.WriteLine(a);
Console.WriteLine(b);
Console.WriteLine(c);
}

Qual a quantidade de vezes que a função foi executada e o valor final
para cada variável.

**Comment by gabriel-hahn** [comment]
Opa @LuizFelipe3, a parte mais complicada dessa função é o `return c + calc(p - 1);` que devemos ter em mente que esse valor acumula a cada nova chamada na função (o valor do **A** vai vir desse _return_).
O **C** é incrementado a cada vez que o `calc` é chamado (3 vezes) ou seja, o valor de **C** fica 3. O `calc` é chamado essa quantidade de vezes porque é quantas chamadas recursivas demorou para o **P** ser menor que 2 (`if (p < 2)`) e retornar um valor fixo 4 e não outra chamada recursiva.
O **B** foi declarado fora do escopo da função e não tem seu valor alterado diretamente, pois o `P` só recebeu o valor de `B` para cálculo, mas continuam sendo coisas separadas, ou seja, **B** continua com o valor **3**.
O **A** que é o mais difícil de calcular, ficou com o valor de **7**: Retorna o valor de **C** somando de cada escopo de chamada da função recursiva, ou seja, 1 e 2 do **C** das primeiras duas vezes que a chamada é recursiva + o retorno da última vez que a função é chamada (**4**) = **7** (isso é um pouco confuso, mas enfim haha).

Como prova real, troquei esse teu código para JS (creio que para você irá trazer o mesmo resultado, só se tiver alguma regra de escopo ou referências diferente do JS) e usei o _console_ para calcular: 

![cálculo](https://user-images.githubusercontent.com/19672684/104727298-d5545100-5713-11eb-9d81-0919150fa7d7.png)

Espero ter ajudado👌🏼
